src/node.js

```diff
@@ -390,6 +391,78 @@
     }
   };
 
+  startup.processPromises = function() {
+    // this should be kept in sync with PromiseRejectEvent in v8.h
+    var kPromiseRejectWithNoHandler = 0;
+    var kPromiseHandlerAddedAfterReject = 1;
+
+    var timers = NativeModule.require('timers');
+    var unhandledPromises = new WeakMap();
+    // A hack to work around uniterable WeakMaps.
+    var unhandledPromiseKeys = [];
+    var fullTurnQueued = false;
+
+    function queueFullTurn() {
+      if (!fullTurnQueued) {
+        fullTurnQueued = true;
+        timers.setImmediate(function() {
+          process.nextTick(afterFullTurn);
+        });
+      }
+    }
+
+    function afterFullTurn() {
+      fullTurnQueued = false;
+      var keys = unhandledPromiseKeys;
+      unhandledPromiseKeys = [];
+      for (var i = 0; i < keys.length; ++i) {
+        var promise = keys[i];
+        var unhandledRejection = unhandledPromises.get(promise);
+        // A handler was added before a full turn happened.
+        if (unhandledRejection === undefined) continue;
+        unhandledRejection.hasBeenNotified = true;
+        var reason = unhandledRejection.reason;
+        if (!process.emit('unhandledRejection', reason, promise)) {
+          // Nobody is listening.
+          // TODO: Take some default action
+          // e.g. logRejection(reason);
+        }
+
+      }
+    }
+
+    function unhandledRejection(promise, value) {
+      unhandledPromises.set(promise, {
+        reason: value,
+        hasBeenNotified: false
+      });
+      unhandledPromiseKeys.push(promise);
+      queueFullTurn();
+    }
+
+    function rejectionHandled(promise) {
+      var unhandledRejection = unhandledPromises.get(promise);
+      if (unhandledRejection !== undefined) {
+        unhandledPromises.delete(promise);
+        // Only fire this event when we have fired a corresponding
+        // 'unhandledRejection' event before it.
+        if (unhandledRejection.hasBeenNotified) {
+          process.emit('rejectionHandled', promise);
+        }
+      }
+    }
+
+    process._setupPromises(function(event, promise, value) {
+      if (event === kPromiseRejectWithNoHandler) {
+        unhandledRejection(promise, value);
+      } else if (event === kPromiseHandlerAddedAfterReject) {
+        rejectionHandled(promise);
+      } else {
+        require('assert').fail('unexpected PromiseRejectEvent');
```

Comment:
Why not `throw new Error('unexpected PromiseRejectEvent')` which is basically what this does?